app/src/main/java/org/wikipedia/search/SearchResultsViewModel.kt

```diff
@@ -0,0 +1,171 @@
+    var languageCode: String? = null
+
+    @OptIn(FlowPreview::class) // TODO: revisit if the debounce method changed.
+    val searchResultsFlow = Pager(PagingConfig(pageSize = batchSize, initialLoadSize = batchSize)) {
+        SearchResultsPagingSource(searchTerm, languageCode, resultsCount, totalResults)
+    }.flow.debounce(delayMillis).map { pagingData ->
+        pagingData.filter { searchResult ->
+            totalResults.find { it.pageTitle.prefixedText == searchResult.pageTitle.prefixedText } == null
+        }.map {
+            totalResults.add(it)
+            it
+        }
+    }.cachedIn(viewModelScope)
+
+    fun clearResults() {
+        resultsCount.clear()
+        totalResults.clear()
+    }
+
+    class SearchResultsPagingSource(
+            val searchTerm: String?,
+            val languageCode: String?,
+            var resultsCount: MutableList<Int>?,
+            var totalResults: MutableList<SearchResult>?
+    ) : PagingSource<MwQueryResponse.Continuation, SearchResult>() {
+
+        private val batchSize = 20
+        private var prefixSearch = true
+
+        override suspend fun load(params: LoadParams<MwQueryResponse.Continuation>): LoadResult<MwQueryResponse.Continuation, SearchResult> {
+            return try {
+                // The default offset is 0 but we send the initial offset from 1 to prevent showing the same talk page from the results.
+                if (searchTerm.isNullOrEmpty() || languageCode.isNullOrEmpty()) {
+                    return LoadResult.Page(emptyList(), null, null)
+                }
+
+                val wikiSite = WikiSite.forLanguageCode(languageCode)
+                var response: MwQueryResponse? = null
+                var readingListSearch = SearchResults()
+                var historySearch = SearchResults()
+                if (prefixSearch) {
+                    if (searchTerm.length > 2) {
+                        readingListSearch = withContext(Dispatchers.IO) {
+                            async {
+                                AppDatabase.instance.readingListPageDao().findPageForSearchQueryInAnyList(searchTerm)
+                            }
+                        }.await()
+
+                        historySearch = withContext(Dispatchers.IO) {
+                            async {
+                                AppDatabase.instance.historyEntryWithImageDao().findHistoryItem(searchTerm)
+                            }
+                        }.await()
+                    }
+                    response = ServiceFactory.get(wikiSite).prefixSearch(searchTerm, params.loadSize, params.key?.gpsoffset)
+                    prefixSearch = false
+                }
+
+                if (response?.query?.pages == null) {
+                    // Prevent using continuation string from prefix search
+                    val continuation = if (params.key?.continuation?.contains("description") == true) null else params.key?.continuation
```

Comment:
When does the continuation string contain `"description"`?

Reply:
Updated the comment.

The `description` string came from the `prefixsearch` API request, which happens in the first round of the `PagingSource`. It will be sent as the `nextKey`, but the next round of `PagingSource` is a `fulltextsearch`, and the `continuation` should be empty.